### PR TITLE
Fixes 14366 upgrade php 72alpha

### PR DESCRIPTION
**Pull request by dynasource**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #14366

### CODE REVIEW

**Review by samdark — APPROVED** [review]
Formatting should be fixed and changelog should be added. Other than that looks good.

**Comment by dynasource** [comment]
> Formatting should be fixed and changelog should be added. Other than that looks good.

Which is formatting?

This is mainly a compatibility patch to make it work with 7.2. Do we need a changelog for that?

File: framework/base/Application.php [hunk 1]
```diff
@@ -300,19 +300,23 @@ protected function bootstrap()
...
+        foreach ($this->bootstrap as $mixed) {
             $component = null;
-            if (is_string($class)) {
-                if ($this->has($class)) {
-                    $component = $this->get($class);
-                } elseif ($this->hasModule($class)) {
-                    $component = $this->getModule($class);
-                } elseif (strpos($class, '\\') === false) {
-                    throw new InvalidConfigException("Unknown bootstrapping component ID: $class");
+            if($mixed instanceof \Closure){
```
Comment (samdark) [inline framework/base/Application.php]: Formatting.

File: framework/base/Application.php [hunk 2]
```diff
@@ -300,19 +300,23 @@ protected function bootstrap()
...
+                if(!$component = call_user_func($mixed, $this)){
```
Comment (samdark) [inline framework/base/Application.php]: Formatting.

File: framework/base/Application.php [hunk 3]
```diff
@@ -300,19 +300,23 @@ protected function bootstrap()
...
+                    continue;
+                }
+            }else if (is_string($mixed)) {
```
Comment (samdark) [inline framework/base/Application.php]: Formatting.

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/base/Application.php [hunk 4]
```diff
@@ -302,17 +302,19 @@ protected function bootstrap()
...
                 }
-            }else if (is_string($mixed)) {
-                if ($this->has($mixed)) {
-                    $component = $this->get($mixed);
-                } elseif ($this->hasModule($mixed)) {
-                    $component = $this->getModule($mixed);
-                } elseif (strpos($mixed, '\\') === false) {
-                    throw new InvalidConfigException("Unknown bootstrapping component ID: $mixed");
+            } else {
+                if (is_string($mixed)) {
```
Comment (samdark) [inline framework/base/Application.php]: This could be `elseif`.

**Review by samdark — APPROVED** [review]

**Review by cebe — COMMENTED** [review]
`Application::$bootstrap` docs need to be updated to list Closure as possible input.

File: framework/base/Application.php [hunk 5]
```diff
@@ -300,19 +300,24 @@ protected function bootstrap()
...
-                if ($this->has($class)) {
-                    $component = $this->get($class);
-                } elseif ($this->hasModule($class)) {
-                    $component = $this->getModule($class);
-                } elseif (strpos($class, '\\') === false) {
-                    throw new InvalidConfigException("Unknown bootstrapping component ID: $class");
+            if ($mixed instanceof \Closure) {
+                if (!$component = call_user_func($mixed, $this)) {
+                    continue;
+                }
```
Comment (cebe) [inline framework/base/Application.php]: in the closure case there should also be a log like in other cases below.
e.g. `Yii::trace('Bootstrap with Closure', __METHOD__);`

**Review by cebe — CHANGES_REQUESTED** [review]
CHANGELOG is also missing.

**Review by cebe — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thanks!

**Pull request merged**